### ISSUE TITLE
Feature request: CMake options for disabling primitives/features

### ISSUE THREAD

**Issue by atafra** [intel/mkl-dnn#573]
# Summary
Adding CMake options for enabling/disabling primitives / major features (e.g. GEMM convolutions, reorders, batch norm) and data types (e.g. int, BF16) to reduce binary size and build time would be very useful for projects that link MKL-DNN statically.

# Problem statement
Libraries that use MKL-DNN and link it statically (e.g. Intel(R) Open Image Denoise) may require only a small subset of the primitives, features, and data types that MKL-DNN supports. Building MKL-DNN with all features results in very large binary sizes and also increases build time significantly.

# Preferred solution
CMake options could be added for enabling/disabling primitives and data types. By default all could be enabled, but projects that need only a subset could choose to enable only the features they need. An example implementation of this concept can be found in Intel(R) Embree: https://github.com/embree/embree/blob/master/CMakeLists.txt (from line 114).

**Comment by vpirogov** [comment]
Thank you for the input. I would consider this a a general request to reduce the library footprint.

Could you please also help me understand how important the size is for OIDN users? Is there a specific target with respect to the binary size you would like to achieve?

**Comment by vpirogov** [comment]
Compile is something nobody asked before. Could you please tell more about this part?

Here's what I see with current master:
```
$ icc --version
icc (ICC) 19.0.5 20190815
Copyright (C) 1985-2019 Intel Corporation.  All rights reserved.
$ CXX=icpc CC=icc time cmake -DDNNL_LIBRARY_TYPE=STATIC -DDNNL_BUILD_TESTS=OFF -DDNNL_BUILD_EXAMPLES=OFF ..
$ time make -j
real    **2m53.495s**
$ ls -lah src/libdnnl.a                            
-rw-r--r-- 1 vpirogov users **84M** Oct 15 14:31 src/libdnnl.a
$: ls -lah examples/cnn-inference-f32-cpp 
-rwxr-xr-x 1 vpirogov users **33M** Oct 15 14:57 examples/cnn-inference-f32-cpp
```

**Comment by atafra** [comment]
This feature request is *not* primarily about general reduction of the library footprint (of course that would be nice too), but the ability to disable unneeded features in MKL-DLL to reduce both library footprint and build time. If many features are disabled, compilation is of course much faster too. It's not a main requirement but a useful side effect. There's no point in compiling source files that will be never used by the application.

**Comment by vpirogov** [comment]
Based on the binary size analysis we did there's no significant value in disabling features from that perspective. Considering the additional validation costs necessary to support new configurations we decided not to proceed with implementation.

**Comment by dzarukin** [comment]
Hi @atafra. If by any chance you are still working with oneDNN, we managed to promote this request with 45f1d2328 on master and functionality will be available in v2.4. Documentation may be found [here](https://github.com/oneapi-src/oneDNN/blob/master/doc/build/build_options.md) (see DNNL_ENABLE_WORKLOAD and DNNL_ENABLE_PRIMITIVE). Thank you.

**Comment by atafra** [comment]
This looks great, I'll have a look soon, thanks!
